packages/runtime/src/next/components/tests/controls/page-control-prop-rendering.tsx

```diff
@@ -86,33 +130,58 @@ export async function testPageControlPropRendering<D extends ControlDefinition>(
       )
     }),
     {
-      type: TestComponentType,
-      label: 'TestComponent',
+      ...testComponentMeta,
       props: {
         propKey: controlDefinition as any,
       },
     },
   )
 
-  // Assert
-  await act(async () =>
-    render(
-      <Testing.ReactProvider runtime={runtime}>
-        <Page snapshot={snapshot} />
-      </Testing.ReactProvider>,
-    ),
+  const testElementTree = (component: ReactNode) => (
+    <Testing.ReactProvider runtime={runtime}>{component}</Testing.ReactProvider>
   )
 
-  if (action) {
-    await act(async () => {
-      await action(screen.getByTestId(testId))
-    })
-  }
-
-  expect(snapshot).toMatchSnapshot('snapshot')
-  expect(propSnapshot(screen.getByTestId(testId))).toMatchSnapshot('resolvedValue')
-
-  if (expectedRenders != null) {
-    expect(Number(screen.getByTestId(renderCountTestId).textContent)).toBe(expectedRenders)
+  if (!isServer()) {
+    const rootElementData: ElementData = Testing.createRootComponent(
+      [elementData, ...rootElements],
+      ROOT_ID,
+    )
+
+    const snapshot = Testing.createMakeswiftPageSnapshot(rootElementData, { locale, cacheData })
+
+    // Assert
+    await act(async () => render(testElementTree(<Page snapshot={snapshot} />)))
+
+    if (action) {
+      await act(async () => {
+        await action(screen.getByTestId(testId))
+      })
+    }
+
+    expect(snapshot).toMatchSnapshot('snapshot')
+    expect(propSnapshot(screen.getByTestId(testId))).toMatchSnapshot('resolvedValue')
+
+    if (expectedRenders != null) {
+      expect(Number(screen.getByTestId(renderCountTestId).textContent)).toBe(expectedRenders)
+    }
+  } else {
+    // test server-side rendering using a component snapshot
```

Comment:
This is the new branch; I opted out for rendering through an embedded component here as it lends itself to cleaner style snapshots (we need to snapshot all of the document's styles, and rendering a single component vs. a whole page means we can avoid snapshotting irrelevant styles).